h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -289,104 +291,105 @@ private static void append(final int[] rightChildren, final int[] leftChildren,
+                    String[] inclusiveLevelsStr = stringToParseInclusiveLevelsFrom.split(",");
+                    for (String level : inclusiveLevelsStr) {
+                        sb.append(node.getDomainValues()[Integer.parseInt(level.replaceAll("\\s", ""))]).append(" ");
+                    }
             }
-            conditionLine += " ])\n";
+            } else {
+                sb.append(" ");
+            }
+            sb.append("]) -> ");
         } else {
-            if (Float.compare(properties._thresholds[index],Float.NaN) == 0) {
+            if (Float.compare(node.getSplitValue(), Float.NaN) == 0) {
                 String sign;
-                if ("R".equals(parentOrigin)) {
+                if ("R".equals(from)) {
                     sign = " is not ";
                 } else {
                     sign = " is ";
                 }
-                conditionLine = "If ( " + properties._features[index] + sign + "NaN )\n";
+                sb.append("If ( ").append(node.getColName()).append(sign).append("NaN )");
             } else {
                 String sign;
-                if ("R".equals(parentOrigin)) {
+                boolean useNan = false;
+                String nanString = " or " + node.getColName() + " is NaN";
+                if ("R".equals(from)) {
                     sign = " < ";
-                    if ("LEFT".equals(properties._nas[index])) {
+                    if (node.getLeftChild().isInclusiveNa()) {
                         useNan = true;
                     }
                 } else {
                     sign = " >= ";
-                    if ("RIGHT".equals(properties._nas[index])) {
+                    if (node.getRightChild().isInclusiveNa()) {
                         useNan = true;
                     }
                 }
-                conditionLine = "If ( " + properties._features[index] + sign + properties._thresholds[index];
+                sb.append("If ( " ).append(node.getColName()).append(sign).append(node.getSplitValue());
                 if (useNan) {
-                    conditionLine += nanString;
+                    sb.append(nanString);
+                  
                 }
-                conditionLine += " )\n";
+                sb.append(" ) -> ");
             }
         }
-        return conditionLine;
+        return sb.toString();
+    }
+
+    private static List<PathResult> findPaths(SharedTreeNode node) {
+        if (node == null)
+            return new ArrayList<>();
+        
+        List<PathResult> result = new ArrayList<>();
+
+        List<PathResult> left_subtree = findPaths(node.getLeftChild());
+        List<PathResult> right_subtree = findPaths(node.getRightChild());
+
+        for (int i = 0; i < left_subtree.size(); i++){
+            PathResult left_result = left_subtree.get(i);
+            PathResult new_result = left_result;
```

Comment:
```suggestion
            PathResult newResult = leftResult;
```

Same sugestion for other varibales down. :)

Reply:
1a3514f